### DISCUSSION TITLE
binding datatable to datagrid ,not use mvvm,how to write.

### DISCUSSION THREAD

**Discussion by qq908382818** [AvaloniaUI/Avalonia#15031] [General]
<-DataGrid x:Name="dg1" AutoGenerateColumns="False">
	<-DataGrid.Columns>
		<-DataGridTextColumn Header="row1"   Binding="{Binding Path=row1}"/>
		<-DataGridTextColumn Header="row2" Binding="{Binding row2}" />
	<-/DataGrid.Columns>
<-/DataGrid>

DataTable dt1 = new();
dt1.Columns.Add("row1", typeof(string));
dt1.Columns.Add("row2", typeof(string));
for(int i = 0; i < 10; i++)
{
    DataRow dr1 = dt1.NewRow();
    dr1["row1"] = "row1_" + i;
    dr1["row2"] = "row2_" + i;
    dt1.Rows.Add(dr1);
}
this.dg1.ItemsSource = dt1.DefaultView;

not work. x:datatype=?? binding =??

**Comment by maxkatz6** [comment]
https://github.com/AvaloniaUI/Avalonia/issues/2781

**Comment by qq908382818** [comment]
not work

**Comment by qq908382818** [comment]
<--DataGrid.Columns
> <--DataGridTextColumn Binding=“{Binding Row.ItemArray[0]”/>
<--/DataGrid.Columns>

  DataGridTextColumn colTmp = new()
  {
      Header = $"第{i}列",
      Binding = new Binding($"Row.ItemArray[{i}]"),
      
  };

how to change to axaml. code

**Reply by maxkatz6** [reply]
If you need dynamically generate columns, you can only do it in C#.

**Reply by qq908382818** [reply]
This is sample code. I actually use fixed columns instead of dynamically generated ones. C# code cannot achieve the effect of setting data templates.
Sorting also becomes ineffective.

**Reply by qq908382818** [reply]
DataGridTemplateColumn colTemplate = new DataGridTemplateColumn();  Can this achieve the customized effect for many columns?

**Reply by maxkatz6** [reply]
> C# code cannot achieve the effect of setting data templates.

You can use FuncDataTemplate - https://docs.avaloniaui.net/docs/concepts/templates/creating-data-templates-in-code.
